Add activity_log table for audit v2

adam-james-v opened the pull request:
Epic: #27661

This PR has a couple elements to it, all meant to help us transition to a unified `activity_log` table where we can write all events that we care about for audit/ 'instance analytics'. This PR focuses on moving `activity` and `view_log` entries first, and does so with:

- migrations to create the `activity_log` table with columns:
```clojure
{:table_id nil
 :database_id nil
 :topic :install
 :details {}
 :model_id nil
 :id 1
 :user_id nil
 :timestamp #t "2023-01-09T20:06:20.832746Z"
 :end_timestamp nil
 :model "install"}
```
- migration to copy activity -> activity_log (past 30 days)
  - drop `:custom_id column`, otherwise things are the same
- migration to copy view_log -> activity_log (past 30 days)
  - `:metadata` is renamed `:details`
  - add `:topic` `(str (:model view-log-entry) "-view")`
- add a multimethod in `metabase.models/activity-log` to use in a model's post-insert for double-write purposes (this is temporary and will be removed once all of the old audit models are removed

A list of questions that I had while working on this, in no particular order:
- can we use proper json columns now that we have newer H2?
- should end_timestamp actually exist? We could model the same kind of idea by instead creating 2 events:
  - :query_start (timestamp then represents the instant a query began)
  - :query_end (timestamp of a different event represents the end)
  - this would let us also create events like :query_failed
- if we have an open json blob in 'details', we might discover that there are certain keys that are essentially expected -> can we anticipate some of those and make them columns on the table directly?

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/27740)
<!-- Reviewable:end -->

adam-james-v committed:
Create migration to add activity_log

Differences between activity_log (new) and activity (old):

- drop `custom_id` from activity, it's unused
- allow `details` to be null
- add nullable `end_timestamp` which can be used for long running events (eg. a slow query)

`details` is a text column in the same way `visualization_settings` is a text column -> it will be storing a json
blob, which will be handled by pre-insert and post-selects, I think.

What I'm not sure about still:
- can we use proper json columns now that we have newer H2?
- should end_timestamp actually exist? We could model the same kind of idea by instead creating 2 events:
  - :query_start (timestamp then represents the instant a query began)
  - :query_end (timestamp of a different event represents the end)
  - this would let us also create events like :query_failed
- if we have an open json blob in 'details', we might discover that there are certain keys that are essentially
expected -> can we anticipate some of those and make them columns on the table directly?

adam-james-v committed:
Add necessary remarks to the migration.

adam-james-v committed:
Add ActivityLog Model, API, and test by duplicating Activity namespaces

This is the first step in moving things over to ActivityLog, which will unify several different tables, but will start with
with Activity and ViewLog.

adam-james-v committed:
Add ActivitLog to cmd copy

adam-james-v committed:
Fix activity-log api endpoint

adam-james-v committed:
Missed these vars

adam-james-v committed:
Only 1 change per changeset

adam-james-v committed:
Wrong fk name fixed

adam-james-v committed:
Make SQL that works with h2

adam-james-v committed:
Pass yarn yaml linter

adam-james-v committed:
Change syntax to work with h2, I think??

adam-james-v committed:
Add method for writing activity and view_log entries to activity_log

This is necessary only for the period of time between making the new activity_log, hooking everything up to it, and
removing the old models/tables. During that dev. time, we may push to master and we don't want to lose and events, so
we double write to make sure we catch them.

adam-james-v committed:
Remove circular dependency by just doing fns in model namespaces directly

Maybe I overdid it with a multimethod? Well, to avoid circular dependency problems with ease, I can just do each
post-insert function directly in each namespace too, not that big a deal.

adam-james-v committed:
Fix some test failures

adam-james-v committed:
Accidentally left this postgres cast of NULL in mysql

noahmoss commented on resources/migrations/000_migrations.yaml:
```diff
@@ -13990,6 +14100,62 @@ databaseChangeLog:
+        - sql:
+            dbms: mysql,mariadb,h2
+            sql: >-
+              INSERT INTO activity_log (table_id, database_id, topic, details, model_id, user_id, timestamp, model)
+              SELECT table_id, database_id, topic, details, model_id, user_id, timestamp, model
+              FROM activity WHERE timestamp > now() - INTERVAL '30' day
+              ORDER BY timestamp ASC;
```
Can't this be used for postgres as well?

noahmoss commented on resources/migrations/000_migrations.yaml:
```diff
@@ -13990,6 +14100,62 @@ databaseChangeLog:
+              FROM activity WHERE timestamp > (now() - CAST('30 days' AS interval))
+              ORDER BY timestamp ASC;
+      rollback: # we do not care to preserve the activity_log from 46->45
+        - sql:
+            sql: >-
+              DELETE FROM activity_log;
```
This doesn't hurt but I don't think it's necessary, since the `createTable` instruction will auto-rollback anyway and delete the entire table.

noahmoss commented on resources/migrations/000_migrations.yaml:
```diff
@@ -13990,6 +14100,62 @@ databaseChangeLog:
+        - sql:
+            dbms: mysql,mariadb,h2
+            sql: >-
+              INSERT INTO activity_log (table_id, database_id, topic, details, model_id, user_id, timestamp, model)
+              SELECT NULL AS table_id,
+                     NULL AS database_id,
+                     CONCAT(model, '-view') AS topic,
+                     metadata AS details,
+                     model_id, user_id, timestamp, model
+              FROM view_log WHERE timestamp > now() - INTERVAL '30' day
+              ORDER BY timestamp ASC;
```
Same question as above, does this not work on postgres?

adam-james-v replied:
Hmm, yea, it does. Not sure why I didn't try that... thanks :)

noahmoss commented on src/metabase/api/activity_log.clj:
```diff
@@ -0,0 +1,275 @@
+(ns metabase.api.activity-log
```
Are this namespace, the model namespace, and the tests all just copied over from the `activity` namespaces verbatim? Or are there changes?

If it's the former, could you just rename the `activity` namespaces in place, and just modify the code directly to work with the new table?

adam-james-v replied:
Yea, that might be a good idea. I can try that out soon.

adam-james-v committed:
Simplify. Turns out things worked in Postgres that I thought didn't.

Nice! Simpler is better :)

adam-james-v committed:
Fix ActivityLog record-activity!

Since custom_id is removed in the new table, we obviously can't try to insert events with that key.

adam-james-v committed:
Fix incorrectly applied foreign key constraint on activity_log table

adam-james-v committed:
Migrations to copy activity/view_log to activity_log

adam-james-v committed:
Attempt to remove Activity (as ActivityLog should be drop-in replacement)

adam-james-v committed:
Use h2x per linter

adam-james-v committed:
Merge branch 'master' into audit-v2-add-activity_log-table
